### PR TITLE
Cheaper stack spilling

### PR DESCRIPTION
**Pull request by bartdesmet**
This addresses issue #11842 by avoiding spilling for certain node types. It's very similar to the strategy taken in Roslyn when spilling the stack for `await` expressions.

Note that this PR leverages the work done in https://github.com/dotnet/corefx/pull/11782 to assert the generated IL. These tests are currently `#ifdef`'d out until integration of that PR, but I've kept them in this review to aid in visualizing the effects of this optimization.

Alas, we didn't have IL-level asserts before, so there's no diff between the original IL produced and the IL after the change. As a point of comparison though, have a look at the sample provided in the issue; its IL now looks like:

```
.method int32 ::lambda_method(class [System.Core]System.Runtime.CompilerServices.Closure,int32)
{
  .maxstack 2
  .locals init (
    [0] int32,
    [1] int32
  )

  .try
  {
    IL_0000: ldarg.1
    IL_0001: stloc.1
    IL_0002: leave      IL_000a
  }
  finally
  {
    IL_0007: endfinally
  }
  IL_0008: ldloc.1
  IL_0009: stloc.0
  IL_000a: ldc.i4.1
  IL_000b: ldloc.0
  IL_000c: add
  IL_000d: ret
}
```

That's a reduction by two instructions, one local slot, and a drop of `.maxstack` by one.

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs [hunk 1]
```diff
@@ -205,6 +207,63 @@ private void EnsureDone()
...
+                {
+                    // Emits ldnull, ldc, initobj, closure constant access, etc.
+                    case ExpressionType.Constant:
+                    case ExpressionType.Default:
+                        return false;
+
+                    // Emits calls to pure RuntimeOps methods with immutable arguments
+                    case ExpressionType.Quote:
+                    case ExpressionType.RuntimeVariables:
+                        return false;
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs]: NB: The rationale for these is that both perform any of the following pure steps:

* Load a constant from the `Closure.Constants` collection; these entries are immutable:
  * A `long[]` for indexes used by `RuntimeVariables` to traverse the parent chain and index into `HoistedLocals`
  * An `Expression` containing an expression tree being `Quote`'d
* Call a runtime helper method in `RuntimeOps`, all of which are pure and are not supposed to throw:
  * Allocation of a `RuntimeVariables` object by simply performing field stores
  * An expression visit for an expression being quoted, at most binding unbound parameters

The `Quote` case could be scrutinized though... The `ExpressionQuoter` will reduce `Extension` nodes which calls into `Reduce` methods. If such a method throws, the timing of the exception changes when spilling is carried out. As far as I know, there's no documentation stating that `Reduce` should not throw, so we may be better off removing the `Quote` case altogether. Let me know what you think.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs [hunk 2]
```diff
@@ -205,6 +207,63 @@ private void EnsureDone()
...
+                        if (field != null)
+                        {
+                            // Emits ldc for the raw value of the field
+                            if (field.IsLiteral)
+                                return false;
+
+                            // For read-only fields we could save the receiver, but
+                            // that's more involved, so we'll just handle static fields
+                            if (field.IsInitOnly && field.IsStatic)
+                                return false;
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs]: I noticed that Roslyn applies the same optimization here. However, I had one thought/question here. According to ECMA-335, section I.8.9.5:

>  Also, under BeforeFieldInit semantics, type initializers are allowed to be executed at or before first access to any static field of that type, at the discretion of the CLI. 

The specification talks about `first access` here, which could be as late as the `ldsfld` instruction. If the type initializer throws, the timing of the `TypeInitializationException` changes in the presence of this optimization. I believe that in practice the type initializer runs earlier, at the point of encountering the first reference to the type in a body of IL code to be JITted. (And of course one can argue that type initializers shall not throw, nor that one should depend on the timing of such exceptions, even if an optimization like this moves the effect to a *later* point in the code.)

**Comment by omariom** [comment]
>  I believe that in practice the type initializer runs earlier, at the point of encountering the first reference to the type in a body of IL code to be JITted.

It has changed recently
https://github.com/dotnet/coreclr/issues/1193

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs (hunk 2 above)
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs]: Copying @omariom's comment to the review discussion in the code:

>> I believe that in practice the type initializer runs earlier, at the point of encountering the first reference to the type in a body of IL code to be JITted.

> It has changed recently
> dotnet/coreclr#1193

Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs]: I did some more experiments to simulate the behavior of deferred execution of the type initializer and the effects of the stack spilling optimization in Roslyn and here. Consider this code:

```csharp
using System;
using System.Threading.Tasks;

class C
{
    static void Main(string[] args)
    {
        Console.WriteLine("Main");
        if (args[0] == "async")
            Async();
        else
            Sync();
    }

    static void Async() => Console.WriteLine(AsyncCore().Result);
    static void Sync()  => Console.WriteLine( SyncCore()       );

    static async Task<string> AsyncCore() => X.Foo + await Task.FromResult(Bar());
    static            string  SyncCore()  => X.Foo +                       Bar();

    static string Bar()
    {
        Console.WriteLine("Bar");
        return "bar";
    }
}

class X
{
    public static readonly string Foo = Get();

    static string Get()
    {
        Console.WriteLine("Foo");
        return "foo";
    }
}
```

The sync and async cases should exhibit the same evaluation order, but the `async` method needs to spill the stack prior to the `await`. The optimization gives `X.Foo` an exemption to spilling into a temporary local.

Results of running this are as follows:

| Flavor | CoreCLR | .NET 4.5 |
|---|---|---|
| Sync | `Main Foo Bar` | `Main Foo Bar` |
| Async | `Main Bar Foo` | `Main Foo Bar` |

**Comment by bartdesmet** [comment]
CC @VSadov

File: src/System.Linq.Expressions/tests/StackSpillerTests.cs [hunk 3]
```diff
@@ -449,6 +450,512 @@ public static void Spill_RequireNoRefArgs_New()
             NotSupported(b);
         }
 
+#if PR_11782 // See https://github.com/dotnet/corefx/pull/11782 for addition of IL decompiler
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/tests/StackSpillerTests.cs]: Once #11782 gets merged in, this `#if` will be removed.

**Comment by bartdesmet** [comment]
Test Innerloop CentOS7.1 Release Build and Test please

**Comment by VSadov** [comment]
Re: BeforeFieldInit 

BeforeFieldInit relaxes guarantees on when static initializers run. While it says "before the first access" and it seems as if it delays initializers as late as possible, the reality is often the opposite - initializers run much earlier, often when type is loaded or first assess to a static member jitted.

It has been always recommended to not rely on timing of static field initialization. If more precise timing is needed, user is supposed to use static constructor. (and in such case compiler would not use  BeforeFieldInit )

Need to think a bit more about moving the Quote. It does perform reducing of Extensions and thus may run 3rd party code.

**Comment by VSadov** [comment]
It feels like treating Quote as pure is a bit risky. Mostly because of reductions that are an extensibility point and can potentially have sideeffects.

Also it feels that, next to the cost of Quoting, avoiding a spill would have much lesser impact.
The benefits in not spilling constants and trivial fields are fairly clear, but for Quotes it does not seem to be so.

I wonder - Is it actually common to spill Quotes? 
Perhaps it would be better to continue spilling?

**Comment by stephentoub** [comment]
Thanks, @VSadov. I'm unclear as to the status of the PR.  Should it be revised? Closed?

**Comment by bartdesmet** [comment]
I agree on the assesment for the `Quote` node case; it changes the timing of reduction side-effects and it's not worth risking a breaking change given that these nodes are rather uncommon in spill positions. I'll remove the case and add a comment to explain the rationale we discussed here.

**Comment by bartdesmet** [comment]
@VSadov, @stephentoub - Please review the change I've made; I think this can be merged once approved.

**Comment by VSadov** [comment]
LGTM

**Pull request merged**